File: frontend/src/services/api/integrations.ts
import type { SlackChannel } from '../../types';
import api from './client';

export interface ProviderStatus {
  connected: boolean;
  expires_at: string | null;
  token_type: string | null;
  site_url: string | null;
  created_at: string | null;
}

export interface AllIntegrationsStatus {
  jira: ProviderStatus;
  google_workspace: ProviderStatus;
  github: ProviderStatus;
  slack: ProviderStatus;
  slack_settings: {
    leadership_channel_id: string | null;
  };
}

export interface SlackTestResult {
  ok: boolean;
  team?: string;
  bot_id?: string;
  error?: string;
}

export const integrationsApi = {
  getStatus: async (): Promise<AllIntegrationsStatus> => {
    const response = await api.get<AllIntegrationsStatus>('/admin/integrations/status');
    return response.data;
  },

  saveGitHubToken: async (token: string): Promise<ProviderStatus> => {
    const response = await api.put<ProviderStatus>('/admin/integrations/github', { token });
    return response.data;
  },

  deleteGitHub: async (): Promise<void> => {
    await api.delete('/admin/integrations/github');
  },

  saveSlackToken: async (token: string): Promise<ProviderStatus> => {
    const response = await api.put<ProviderStatus>('/admin/integrations/slack', { token });
    return response.data;
  },

  deleteSlack: async (): Promise<void> => {
    await api.delete('/admin/integrations/slack');
  },

  updateSlackSettings: async (data: {
    leadership_channel_id?: string;
  }): Promise<{ leadership_channel_id: string | null }> => {
    const response = await api.put('/admin/integrations/slack/settings', data);
    return response.data;
  },

  getSlackChannels: async (): Promise<SlackChannel[]> => {
    const response = await api.get<SlackChannel[]>('/admin/integrations/slack/channels');
    return response.data;
  },

  testSlackConnection: async (): Promise<SlackTestResult> => {
    const response = await api.post<SlackTestResult>('/admin/integrations/slack/test');
    return response.data;
  },
};

File: backend/app/modules/scorecard/api/schemas/integrations.py
"""Integration admin API schemas."""

from pydantic import BaseModel, Field


class ProviderStatus(BaseModel):
    """Status of an individual integration provider."""

    connected: bool
    expires_at: str | None = None
    token_type: str | None = None
    site_url: str | None = None
    created_at: str | None = None


class AllIntegrationsStatus(BaseModel):
    """Status of all integration providers."""

    jira: ProviderStatus
    google_workspace: ProviderStatus
    github: ProviderStatus
    slack: ProviderStatus
    slack_settings: dict[str, str | None] = {}


class GitHubTokenInput(BaseModel):
    """Input for saving a GitHub Personal Access Token."""

    token: str = Field(..., min_length=1, description="GitHub Personal Access Token")


class SlackTokenInput(BaseModel):
    """Input for saving a Slack bot token."""

    token: str = Field(..., min_length=1, description="Slack bot token (xoxb-...)")


class SlackSettingsUpdate(BaseModel):
    """Update Slack integration settings."""

    leadership_channel_id: str | None = Field(None, max_length=50)
